### DISCUSSION TITLE
Kiali doesn't see istio-ingressgateway installed in separate namespace

### DISCUSSION THREAD

**Discussion by Bytamine** [kiali/kiali#3759] [Q&A] [answered]
**Describe the bug**
I have istio 1.9 installed in Azure AKS and trying to install kiali-operator 1.31.0 via helm.
Istio is installed into istio-system namespace. Istio Ingress is in istio-ingress. Kiali-operator is installed into kiali-operator. And Kiali CR is in istio-system.
Here is part of CR yaml:
```
    istio_component_namespaces:
      istio-ingressgateway: "istio-ingress"
    istio_namespace: "istio-system"
```
But when I'm looking at Istio Component Status in Kiali UI, I'm getting "istio-ingressgateway Not found". There are no error in kiali logs and I'm out of ideas where to look. Is there some config setting I missed?

**Versions used**
Kiali: 1.31.0
Istio: 1.9.0
Kubernetes flavour and version: AKS 1.17.13

**Comment by Bytamine** [comment]
Full kiali CR
```
annotations:
  ansible.operator-sdk/verbosity: "1"
apiVersion: kiali.io/v1alpha1
kind: Kiali
metadata:
  annotations:
    meta.helm.sh/release-name: kiali-operator
    meta.helm.sh/release-namespace: kiali-operator
  creationTimestamp: "2021-03-09T18:06:19Z"
  finalizers:
  - finalizer.kiali
  generation: 2
  labels:
    app: kiali-operator
    app.kubernetes.io/instance: kiali-operator
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: kiali-operator
    app.kubernetes.io/part-of: kiali-operator
    app.kubernetes.io/version: v1.31.0
    helm.sh/chart: kiali-operator-1.31.0
    version: v1.31.0
  name: kiali
  namespace: istio-system
  resourceVersion: "107178885"
  selfLink: /apis/kiali.io/v1alpha1/namespaces/istio-system/kialis/kiali
  uid: 688cd59f-fd81-48b8-a013-6ac2a2209bf6
spec:
  deployment:
    accessible_namespaces:
    - '**'
    node_selector:
      agentpool: istio
    override_ingress_yaml:
      spec:
        rules:
        - host: kiali.staging.domain.com
          http:
            paths:
            - backend:
                serviceName: kiali
                servicePort: 20001
              path: /
  external_services:
    grafana:
      in_cluster_url: http://prometheus-operator-grafana.kube-monitoring
    prometheus:
      url: http://prometheus-operator-prometheus.kube-monitoring:9090
  istio_component_namespaces:
    istio-ingressgateway: istio-ingress
  istio_namespace: istio-system
status:
  accessibleNamespaces: '**'
  conditions:
  - ansibleResult:
      changed: 20
      completion: 2021-03-09T18:15:20.709623
      failures: 0
      ok: 159
      skipped: 77
    lastTransitionTime: "2021-03-09T18:06:25Z"
    message: Awaiting next reconciliation
    reason: Successful
    status: "True"
    type: Running
  progress:
    duration: "0:00:45"
    message: 17. Finished
   ```

**Comment by jmazzitelli** [comment]
We really need to revisit the whole "istio_component_namespaces" config. That was used [by Kiali for a time when Istio's installer was trying to provide ways to split up the control plane across multiple namespaces.](https://github.com/kiali/kiali/issues/1315) They have since moved away from that in their installer. Kiali still "supports" the   istio_component_namespaces settings, but not in the way you are expecting. What you have is not going to do what you think it is going to do. [Here you can see where the operator uses this istio_component_namespaces](https://github.com/kiali/kiali-operator/blob/d8ee514bffde9c3eb1fac4d0a430a63da1b7c40e/roles/default/kiali-deploy/tasks/main.yml#L195-L223) and it does not look for the ingress gateway as a component namespace. That dict config `istio_component_namespaces` is only used to tell the operator where these components are: grafana, jaeger (aka tracing), istiod, and prometheus. It is not for configuring a separate location for an ingress gateway (because the operator doesn't care where one is).

I'm going to convert this to a github "discussion" because it seems to be simply a misconfiguration somewhere. We can further discuss - I'm not sure who the correct person is for this - perhaps @xeviknal

**Reply by jmazzitelli** [reply]
cc @jshaughn we should probably re-visit this once again and see if we can remove that setting. It is just adding confusion for no benefit that I can see now.

**Reply by jshaughn** [reply]
I don't think we can just remove it but I agree that we need to revisit and see what makes more sense.

**Reply by xeviknal** [reply]
It's late now for me, I couldn't read the whole threat. I'll do it tomorrow.
I find this `istio_component_namespaces` config pretty confusing name-wise. This is not the config I am using for computing the masthead health status.

+99 to simplify the process and config in this area to a support the whole set of ways to install istio.

**Comment by xeviknal** [comment, marked as answer]
Hi @andreybleme,

you should modify this bit of the CR: https://github.com/kiali/kiali-operator/blob/master/deploy/kiali/kiali_cr.yaml#L501
```
    istio:
      component_status:
        enabled: true
        components:
        - app_label: istiod
          is_core: true
        - app_label: istio-ingressgateway
          is_core: true
          namespace: istio-ingress
        - app_label: istio-egressgateway
          is_core: false
```


As far as I can remember, `istio_component_namespaces` config doesn't belong to the istio component status feature. I can't remember what it is used for.

In our FAQ, this should be documented: https://kiali.io/documentation/latest/faq/#component-not-found

Let me know if you need a further support. Happy to help clarifying and improving the process.

**Reply by Bytamine** [reply]
Thank you @xeviknal! How can I miss it?
I suggest to add namespaces to example (as Istio guides recommends to deploy *gresses to different namespaces), it will clarify this setup a lot.
```
#    istio:
#      component_status:
#        enabled: true
#        components:
#        - app_label: istiod
#          is_core: true
#        - app_label: istio-ingressgateway
#          is_core: true
#          namespace: istio-ingress
#        - app_label: istio-egressgateway
#          is_core: false
#          namespace: istio-egress
```

**Reply by jmazzitelli** [reply]
> I suggest to add namespaces to example

That example CR explicitly [sets the comments to be the actual default values](https://github.com/kiali/kiali-operator/blob/v1.31.0/deploy/kiali/kiali_cr.yaml#L16-L18). We do this as documentation to tell the reader what the defaults are for all settings in the CR. So it would not be appropriate to show "istio-ingress" and "istio-egress" as the default values for those namespaces because that is not their defaults (the defaults for those settings an empty, nil, value which falls back to the name of the control plane namespace [as documented here in that example file](https://github.com/kiali/kiali-operator/blob/v1.31.0/deploy/kiali/kiali_cr.yaml#L501)).

We could add some more description to that "namespace" documentation. Something like:

```
#     namespace: The namespace where the component is installed in. It defaults to the 'istio_namespace' setting.
                 Note that the Istio documentation suggests you install the ingress and egress to different namespaces,
                 so you most likely will want to explicitly set this namespace value for the ingress and egress components.
                 For example, something like:
                 - app_label: istio-ingressgateway
                   is_core:   false
                   namespace: <the namespace where your ingress is deployed>
```
